### PR TITLE
Fix build and make join table sort stable by using DFS

### PR DESCRIPTION
**Pull request by diosmosis**
Changes:

* Fix original format tests that fail on mysqli due to it returning ints by casting Site properties to int.
* Make sure table sort in join tables is stable by using DFS instead of weighting tables. Join dependencies are parsed/assumed and dependent table/table-aliases get ordered first.
* Couple other minor test fixes.

### CODE REVIEW

**Comment by diosmosis** [comment]
Refs #13025

**Review by sgiehl — COMMENTED** [review]
Code changes look good to me and tests seem to pass. Would maybe be good if @tsteur would have another look

File: core/DataAccess/LogQueryBuilder/JoinTables.php [hunk 1]
```diff
@@ -137,5 +132,147 @@ private function checkTableCanBeUsedForSegmentation($tableName)
...
+    private function parseDependencies(array $tables)
+    {
+        // NOTE: joins can be specified explicitly as arrays w/ 'joinOn' keys or implicitly as table names. when
+        // table names are used, the joins dependencies are assumed based on how we want to order those joins.
+        // the below table list the possible dependencies of each table, and is specifically designed to enforce
+        // the following order:
+        // log_link_visit_action, log_action, log_visit, log_conversion, log_conversion_item
+        // which means if an array is supplied where log_visit comes before log_link_visitAction, it will
+        // be moved to after it.
+        static $implicitTableDependencies = [
```
Comment (tsteur) [inline core/DataAccess/LogQueryBuilder/JoinTables.php]: is that needed to be static?

**Comment by tsteur** [comment]
It's a complex PR/topic and don't understand all the code / hard to review so focused on looking at tests. Ran the system tests for most premium features and they seem to pass. Also checked custom reports generates similar queries that still work (some SQLs changed but are still valid). As nothing seems to break I reckon it's good to merge  (in 3.8).

File: core/DataAccess/LogQueryBuilder/JoinTables.php (hunk 1 above)
Comment (diosmosis) [reply core/DataAccess/LogQueryBuilder/JoinTables.php]: Not really, it's just a constant. I can remove the static modifier.

File: core/Site.php [hunk 2]
```diff
@@ -82,6 +82,13 @@ public function __construct($idsite)
         self::setSiteFromArray($this->id, $site);
 
         $this->site = $site;
+
+        // for serialized format to be predictable across php/mysql/pdo/mysqli versions, make sure the int props stay ints
+        foreach ($this->site as $key => $value) {
+            if (is_numeric($value)) {
```
Comment (tsteur) [inline core/Site.php]: what about floats etc? It'd be actually better to really specify the properties that should be converted (like we do in other plugins) and not something generic

**Comment by tsteur** [comment]
fyi we will need to test this with all premium features again, and especially custom reports to make sure the generated queries are still correct.

**Comment by tsteur** [comment]
ran the custom reports tests  @diosmosis and looks all good 👍

File: core/Site.php (hunk 2 above)
Comment (diosmosis) [reply core/Site.php]: changed

**Pull request merged**